packages/synapse-react-client/src/components/DataGrid/DataGrid.tsx

```diff
@@ -0,0 +1,695 @@
+import { useState, useRef, useEffect, useMemo } from 'react'
+import {
+  DataSheetGrid,
+  keyColumn,
+  Column,
+  createTextColumn,
+} from 'react-datasheet-grid'
+import 'react-datasheet-grid/dist/style.css'
+import '../../style/components/_data-grid-extra.scss'
+import { useSynapseContext } from '@/utils/context/SynapseContext'
+import { SynapseClient, TableQuery } from '@sage-bionetworks/synapse-client'
+import { useCreateGridSession } from './useCreateGridSession'
+import {
+  CreateGridRequest,
+  PostRepoV1GridSessionIdReplicaRequest,
+} from '@sage-bionetworks/synapse-client'
+import { Model, VecApi } from 'json-joy/lib/json-crdt'
+import { encode, decode } from 'json-joy/lib/json-crdt-patch/codec/compact'
+import { konst } from 'json-joy/lib/json-crdt-patch'
+import throttle from 'lodash.throttle'
+import { Encoder as VerboseEncoder } from 'json-joy/lib/json-crdt/codec/structural/verbose/Encoder'
+
+const verboseEncoder = new VerboseEncoder()
+
+const DataGrid = () => {
+  const synapseClient = useSynapseContext().synapseClient
+
+  // Grid session state
+  const [sessionId, setSessionId] = useState<string>('')
+  const [replicaId, setReplicaId] = useState<number | null>(null)
+  const [presignedUrl, setPresignedUrl] = useState<string>('')
+  const [gridSql, setGridSql] = useState<string>('')
+  const inputRef = useRef<HTMLInputElement>(null)
+  const webSocketRef = useRef<WebSocket | null>(null)
+
+  // WebSocket state
+  const [isConnected, setIsConnected] = useState<boolean>(false)
+  const [connectionStatus, setConnectionStatus] =
+    useState<string>('Disconnected')
+  const [messages, setMessages] = useState<string[]>([])
+
+  // CRDT state
+  const sequenceNumberRef = useRef<number>(1)
+  const modelRef = useRef<Model | null>(null)
+  interface ModelSnapshot {
+    columnNames: string[]
+    columnOrder: number[]
+    rows: string[]
+  }
+  interface jsonJoyMessage {
+    sequenceNumber: number
+    methodName: string
+    payload?: string
+  }
+  const modelSnapshotRef = useRef<ModelSnapshot>({
+    columnNames: [],
+    columnOrder: [],
+    rows: [],
+  })
+  const getModel = () => modelRef.current
+  const setModel = (newModel: Model) => {
+    if (newModel.view() !== modelRef.current?.view()) {
+      modelRef.current = newModel
+      modelSnapshotRef.current = modelRef.current.api.getSnapshot()
```

Comment:
The model itself will change with every patch, but the snapshot contains the essential information for rendering the grid - `columnNames`, `columnOrder`, and `rows`. So the snapshot should be computed each model update, but re-renders should only happen when the snapshot changes.